### DISCUSSION TITLE
Changes cxf Client URL only in Test

### DISCUSSION THREAD

**Discussion by arolfes** [quarkiverse/quarkus-cxf#319] [Q&A] [answered]
Hi,

I wanna use wiremockserver with dynamic port to simulate different responses and validate the incomig requests .

Therefore I need to change the port where the injected Cxf Client sends the request. 

Is it possible? 
If yes, can you give me a hint how to change the port at test runtime? Probably inside BeforeAll method ?

**Comment by shumonsharif** [comment]
Hi @arolfes You can modify the URL dynamically at runtime using something like:

```java
    @CXFClient
    CalculatorSoap calculatorSoap;

    @BeforeEach
    void setup() {
        ((BindingProvider)calculatorSoap).getRequestContext()
                      .put(BindingProvider.ENDPOINT_ADDRESS_PROPERTY, "http://localhost:9999/calculator.asmx");
    }
```

**Comment by famod** [comment, marked as answer]
How do you start the wiremock server? `@QuarkusTestResource` should just work here, see https://quarkus.io/guides/getting-started-testing#quarkus-test-resource

**Comment by arolfes** [comment]
Hello @shumonsharif , 

thank you for the solution. its working fine. I didn't know that it is still possible to use the _old fashion_ way to change the URL for an endpoint. but when I saw your solution I remembered it from previous cxf projects :-) thx a lot.

Hello @famod ,
yes you are right I followed the tutorial for qarkus-test-resource.

_Late Night Update_: Following the tutorial correctly I don't need to change the URL with extra code. Thank you for making it clear. 

so my code looks like this:

### Consumer with injected client
```kotlin
@ApplicationScoped
class MyConsumer(
    @CXFClient("soapService")
    val soapClient: SoapServicePortType
) {
// some functions...
}
```

### ConsumerTest
```kotlin
class MyConsumerTest {

    @InjectWireMock
    lateinit var wireMockServer: WireMockServer

    @Inject
    lateinit var  consumer: MyConsumer

    @BeforeEach
    fun beforeEach() {
        wireMockServer.resetAll()
    }
// some test functions
```

### Wiremock Resource
```kotlin
class WireMockResource : QuarkusTestResourceLifecycleManager {

    val wireMockServer =
        WireMockServer(WireMockConfiguration.wireMockConfig().dynamicPort().httpsPort(0))

    override fun start(): MutableMap<String, String> {
        wireMockServer.start()

        return mutableMapOf("quarkus.cxf.client.soapService.client-endpoint-url" to "http://localhost:${wireMockServer.port()}/${MyConsumer.SERVICE_ENDPOINT}")
    }

    override fun stop() {
        wireMockServer.stop()
    }

    override fun inject(testInjector: TestInjector) {
        testInjector.injectIntoFields(
            wireMockServer,
            AnnotatedAndMatchesType(InjectWireMock::class.java, WireMockServer::class.java)
        )
    }
}

@Target(AnnotationTarget.FIELD, AnnotationTarget.VALUE_PARAMETER)
@Retention
annotation class InjectWireMock
```

**Comment by arolfes** [comment]
And by the way

your extension is easy to use. Good Job.

**Comment by arolfes** [comment]
Hi @shumonsharif , @famod 

I marked @famod reply as answer because for my test purpose it's the better solution.
The solution from @shumonsharif works also. I tested it first :-)  . 
But I believe using the `QuarkusTestResourceLifecycleManager` is more the quarkus way.

**Reply by shumonsharif** [reply]
Perfect, I also learned something new here ... thanks @arolfes and @famod !  `QuarkusTestResourceLifecycleManager` is fairly new to me as well, and definitely looks like the right way to go.  FWIW, I think the docs for `QuarkusTestResourceLifecycleManager` could use a slight improvement ... it's not obvious (at least it wasn't to me from the docs) that the property values can be overridden in the start method.

**Reply by famod** [reply]
> it's not obvious (at least it wasn't to me from the docs) that the property values can be overridden in the start method

PRs are welcome! 😄

One thing to keep in mind with `@QuarkusTestResource` is that it's "global". `QuarkusTestResource.restrictToAnnotatedClass` can be used to isolate it, but then there will be a Quarkus restart which can be problematic if you have many tests.
But there will at least be automatic test class ordering in Quarkus 2.6 to mitigate the problem, see https://github.com/quarkusio/quarkus/pull/21748 and https://github.com/quarkusio/quarkus/pull/21610.
